File: python/check_if_two_string_arrays_are_equivalent.py
class Solution:
    def arrayStringsAreEqual(self, word1: list[str], word2: list[str]) -> bool:
        """time complexity: O(n)"""
        w1: int = 0
        w2: int = 0
        i: int = 0
        j: int = 0
        while w1 < len(word1) and w2 < len(word2):
            if word1[w1][i] != word2[w2][j]:
                return False
            i += 1
            j += 1
            if i == len(word1[w1]):
                i = 0
                w1 += 1
            if j == len(word2[w2]):
                j = 0
                w2 += 1
        if w1 != len(word1) or w2 != len(word2):
            return False
        return True


def test_arrayStringsAreEqual_case_1():
    # arrange
    word1: list[str] = ["ab", "c"]
    word2: list[str] = ["a", "bc"]
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.arrayStringsAreEqual(word1, word2)

    # assert
    assert expected == actual


def test_arrayStringsAreEqual_case_2():
    # arrange
    word1: list[str] = ["a", "cb"]
    word2: list[str] = ["ab", "c"]
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.arrayStringsAreEqual(word1, word2)

    # assert
    assert expected == actual


def test_arrayStringsAreEqual_case_3():
    # arrange
    word1: list[str] = ["abc", "d", "defg"]
    word2: list[str] = ["abcddefg"]
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.arrayStringsAreEqual(word1, word2)

    # assert
    assert expected == actual

File: rust/src/code/check_if_two_string_arrays_are_equivalent.rs
struct Solution;

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn array_strings_are_equal(word1: Vec<String>, word2: Vec<String>) -> bool {
        let mut new_word1: Vec<Vec<char>> = Vec::new();
        let mut new_word2: Vec<Vec<char>> = Vec::new();
        for word in word1 {
            new_word1.push(word.chars().collect::<Vec<char>>());
        }
        for word in word2 {
            new_word2.push(word.chars().collect::<Vec<char>>());
        }
        let mut w1: usize = 0;
        let mut w2: usize = 0;
        let mut i: usize = 0;
        let mut j: usize = 0;
        while w1 < new_word1.len() && w2 < new_word2.len() {
            if new_word1[w1][i] != new_word2[w2][j] {
                return false;
            }
            i += 1;
            j += 1;
            if i == new_word1[w1].len() {
                w1 += 1;
                i = 0;
            }
            if j == new_word2[w2].len() {
                w2 += 1;
                j = 0;
            }
        }
        if w1 != new_word1.len() || w2 != new_word2.len() {
            return false;
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_array_strings_are_equal_case_1() {
        // arrange
        let word1: Vec<String> = vec!["ab".to_string(), "c".to_string()];
        let word2: Vec<String> = vec!["a".to_string(), "bc".to_string()];
        let expected: bool = true;

        // act
        let actual = Solution::array_strings_are_equal(word1, word2);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_array_strings_are_equal_case_2() {
        // arrange
        let word1: Vec<String> = vec!["a".to_string(), "cb".to_string()];
        let word2: Vec<String> = vec!["ab".to_string(), "c".to_string()];
        let expected: bool = false;

        // act
        let actual = Solution::array_strings_are_equal(word1, word2);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_array_strings_are_equal_case_3() {
        // arrange
        let word1: Vec<String> = vec!["abc".to_string(), "d".to_string(), "defg".to_string()];
        let word2: Vec<String> = vec!["abcddefg".to_string()];
        let expected: bool = true;

        // act
        let actual = Solution::array_strings_are_equal(word1, word2);

        // assert
        assert_eq!(actual, expected);
    }
}
